Ke/tacos

kierse opened the pull request.

kierse commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/MainActivity.kt:
```diff
@@ -0,0 +1,51 @@
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+
+    val circuitConfig: CircuitConfig =
+      CircuitConfig.Builder()
+        .addPresenterFactory(buildPresenterFactory())
+        .addUiFactory(OrderTacosUiFactory())
+        .build()
```
Note: I didn't bother to use Dagger here. It felt like overkill

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/repository/IngredientsRepository.kt:
```diff
@@ -0,0 +1,117 @@
+
+object IngredientsRepositoryImpl : IngredientsRepository {
+  override suspend fun getFillings(): ImmutableList<Ingredient> {
+    delay(Random.nextLong(0, 5000))
```
Why the random delay?

kierse replied:
There are loading states for most steps. I think I initially added this so they were visible

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/model/Ingredient.kt:
```diff
@@ -0,0 +1,18 @@
+
+typealias Cents = Int
+
+@Parcelize
```
Why does it need to be parcelable?

kierse replied:
Nope. Removed in [06b1250](https://github.com/slackhq/circuit/pull/496/commits/06b1250f27c1e130a3063695e3ee2086a0bc8ed4)

ZacSweers commented on samples/tacos/build.gradle.kts:
```diff
@@ -0,0 +1,66 @@
+  namespace = "com.slack.circuit.tacos"
+
+  defaultConfig {
+    minSdk = 33
+    testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
```
I think both of these are handled automatically by our setup in the root build file

kierse replied:
Looks like I can remove `testInstrumentationRunner`. Removing `minSdk` results in an error:
```
In order to use core library desugaring, please enable multidex.
```
We appear to set a default `minSdk` for library and test but not application

ZacSweers commented on samples/tacos/build.gradle.kts:
```diff
@@ -0,0 +1,66 @@
+  .configureEach {
+    compilerOptions {
+      freeCompilerArgs.addAll(
+        "-opt-in=androidx.compose.material.ExperimentalMaterialApi",
```
Do we use M2 anymore anywhere?

kierse replied:
Good catch; it can be removed ([24f2538](https://github.com/slackhq/circuit/pull/496/commits/24f2538f17345b56772263b114aa1b5bc4aa5935)). I started by copying the star app build file and mustn't have looked that closely.

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,360 @@
+  ) : CircuitUiState
+
+  sealed interface Event : CircuitUiEvent {
+    val value: Int
```
nit: this is a little vague for the name, can we improve it?

kierse replied:
Good point. Addressed in [42d6b45](https://github.com/slackhq/circuit/pull/496/commits/42d6b45bc3fb920a2854ece5b86a85beb9b22171)

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,360 @@
+  val ingredientsCost: BigDecimal = BigDecimal.ZERO
+) : Parcelable
+
+private val orderSteps =
```
nit: `ORDER_STEPS`

kierse replied:
I initially went with that, but it violates our lint rules

ZacSweers commented on samples/tacos/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,17 @@
+    <application
+        android:supportsRtl="true"
+        android:theme="@style/TacoTheme">
+        <activity
```
nit: any reason not to use the configChanges param to take full control here?

kierse replied:
I briefly considered going that route but didn't explore it. I don't see any reason not to do that.

I updated to use `configChanges` and removed all uses of `rememberSaveable` and `Parcelable`: [06b1250](https://github.com/slackhq/circuit/pull/496/commits/06b1250f27c1e130a3063695e3ee2086a0bc8ed4)

ZacSweers reviewed (commented):
This looks sick! Started an initial pass with some nits, but want to pause here to ask for a couple early updates to simplify review

- Can you add a README to this sample and flesh it out with details about the structure? Aside from it helping with code review here, it will also be needed for actually helping others understand this sample. There's very little docs in the sample code itself, so it's hard to grok much from the outside without added context.
- I see parcelable used in multiple places that don't seem necessary, can we simplify these? Worried about them being sort of misdirects leading people to think they belong in `Screen`s.
  - Can the sample take control of config changes and avoid needing rememberSaveable/rememberRetained? I suspect these may be why you're using parcelable, but for the sake of keeping the sample simpler I think it's better if we avoid this.

kierse commented:
In addition to the README, I went through and added comments where I felt it might clarify something. I also renamed a few properties and objects to more clearly indicate their purpose

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/MainActivity.kt:
```diff
@@ -0,0 +1,56 @@
+
+    setContent {
+      TacoTheme {
+        rememberSystemUiController().run { setStatusBarColor(MaterialTheme.colorScheme.background) }
```
```suggestion
        rememberSystemUiController().setStatusBarColor(MaterialTheme.colorScheme.background)
```

kierse replied:
I don't think so, but may be unaware of something. This is using Accompanist to set the system status bar colour, not the app top bar colour. I don't think the scaffold can set that

ZacSweers replied:
the top app bar in M3 can fill into the status bar and then all we should normally have to do is set the status bar icons to be light or dark

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+  ) : CircuitUiState
+
+  sealed interface Event : CircuitUiEvent {
+    val indexModifier: Int
+      get() = 0
```
Looks like this is always defined?

kierse replied:
Yup, removed. https://github.com/slackhq/circuit/pull/496/commits/b723dc52ca7f6a984b0de64d4092fd59f209db32

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+
+@Immutable
+data class OrderDetails(
+  val filling: Ingredient = Ingredient(""),
+  val toppings: Set<Ingredient> = persistentSetOf(),
+  val baseCost: BigDecimal = BigDecimal("9.99"), // Default taco price
+  val ingredientsCost: BigDecimal = BigDecimal.ZERO
```
Why all the defaults?

kierse replied:
Those values represent the default state for an order. The user builds on these defaults when interacting with the app. `baseCost` is _always_ 9.99 so I guess it doesn't need to be a constructor arg. Removed as part of https://github.com/slackhq/circuit/pull/496/commits/53e915711ad82900bcfe50764d6511b5f1987d26

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+data class OrderDetails(
+  val filling: Ingredient = Ingredient(""),
+  val toppings: Set<Ingredient> = persistentSetOf(),
+  val baseCost: BigDecimal = BigDecimal("9.99"), // Default taco price
+  val ingredientsCost: BigDecimal = BigDecimal.ZERO
```
`BigDecimal`/floating point in general is an anti-pattern for currency. Let's just use an int that represents cents, then make the render string format it

kierse replied:
Yeah...I started out using cents (ie. an `Int`) then switched to `BigDecimal` to simplify conversion to dollars in the UI.

Reintroduced `Cents` typealias and added a extension to handle conversion to dollars. https://github.com/slackhq/circuit/pull/496/commits/53e915711ad82900bcfe50764d6511b5f1987d26

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+
+@Immutable
+data class OrderDetails(
+  val filling: Ingredient = Ingredient(""),
+  val toppings: Set<Ingredient> = persistentSetOf(),
```
Seems this should be two different ones? `Filling` and `Topping`?

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+  persistentListOf(FillingsOrderStep, ToppingsOrderStep, ConfirmationOrderStep, SummaryOrderStep)
+
+internal class OrderTacosPresenter(
+  private val fillingsProducer: OrderStep.StateProducer<FillingsOrderStep.State>,
+  private val toppingsProducer: OrderStep.StateProducer<ToppingsOrderStep.State>,
+  private val confirmationProducer: OrderStep.StateProducer<ConfirmationOrderStep.OrderState>,
+  private val summaryProducer: OrderStep.StateProducer<SummaryOrderStep.SummaryState>,
```
This isn't really doing a composite presenter since these aren't presenters. Can we make them presenters instead for the sake of this example?

kierse replied:
> This isn't really doing a composite presenter since these aren't presenters

I disagree here. In my view this _is_ a composite presenter. It's composed of other objects that generate the state it needs before it can emit its own state.

When I started I assumed I would use Circuit Presenters here, however, I quickly realized that wasn't necessary or really appropriate.

These state producers aren't standalone circuits. They aren't displayed elsewhere in the application and are completely unknown to the framework. While I _could_ have them implement Circuit's Presenter interface, there's no real benefit. Doing so would further complicate the code as I'd have to create a Presenter factory for each one and inject that into `OrderTacosPresenter` so that it can provide an event sink when calling them.

ZacSweers replied:
hmmm. I don't disagree, but I do also think it's sort of acting as a different kind of sample rather than being a literal composite presenter. Maybe we should have another example for that separately with a simpler case, like an index/detail flow where it's a composite presenter in tablet and multi-screen in phone. In short - I think it's more of a _complex_ (but not complicated!) presenter but not strictly a _composite_ presenter (i.e. literally composed of multiple reusable nested presenters) in the way I thought we were talking about before. I think both are useful 👍.

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+  private val confirmationProducer: OrderStep.StateProducer<ConfirmationOrderStep.OrderState>,
+  private val summaryProducer: OrderStep.StateProducer<SummaryOrderStep.SummaryState>,
+) : Presenter<OrderTacosScreen.State> {
+  @Composable override fun present(): OrderTacosScreen.State = presentInternal()
+
+  @Composable
+  internal fun presentInternal(
+    initialStep: OrderStep = FillingsOrderStep,
+    initialOrderDetails: OrderDetails = OrderDetails(),
+  ): OrderTacosScreen.State {
```
it feels like an anti-pattern for circuit that we're not testing from the `present()` function alone 🤔. Could we do this another way? Maybe make an overloaded constructor that takes these initial args. I think we actually _have_ to do that, because this function can be re-called multiple times so `initial` isn't the right way to think of it here

kierse replied:
Sure, I can look at overloading the constructor and see where that gets me

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+  }
+
+  @Composable
+  private fun produceState(
```
I would define each of these variables directly rather than this function. It's also a bit of a confusing name given the existence of `produceState` in compose itself for suspend functions

kierse replied:
> I would define each of these variables directly rather than this function

I extracted them to a separate function to improve readability. Calling these directly just makes the call site a lot more complicated and messy in my opinion.

> It's also a bit of a confusing name given the existence of produceState in compose itself for suspend functions

Hmm good point. Renamed to `produceOrderStepState`. https://github.com/slackhq/circuit/pull/496/commits/0d0489752bfcb54a7c5591aa271c5cef10531a03

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+  onNavEvent: (OrderStep) -> Unit
+) {
+  val newIndex = currentStep.index + navEvent.indexModifier
+  onNavEvent(orderSteps.getOrElse(newIndex) { currentStep })
```
this seems strange to me, is there ever a case where we have an unmatched index?

kierse replied:
No, not unless there's a programming error. When determining the appropriate next `OrderStep`, I'm calculating a `List` index. I added `getOrElse` simply to avoid the possibility of a out of bounds exception.

ZacSweers replied:
I kind of think we should error instead. Something something offensive programming rather than defensive. But my LOGAF is low

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -0,0 +1,356 @@
+) {
+  if (!visible) return
+
+  val tintColour =
+    when (enabled) {
+      true -> MaterialTheme.colorScheme.onSurface
+      false -> MaterialTheme.colorScheme.outline
+    }
```
I think this means something else is wrong as you shouldn't need to do this manually

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/step/CommonUi.kt:
```diff
@@ -0,0 +1,128 @@
+
+@Preview
+@Composable
+internal fun PreviewOrderIngredient() {
```
previews should be private, not internal

also minor nit but I think it's better to co-locate them with the thing they're previewing rather than putting all of them at the bottom

ZacSweers replied:
Nope not recently! We even lint against it :) : https://github.com/slackhq/compose-lints/blob/main/compose-lint-checks/src/main/java/slack/lint/compose/PreviewPublicDetector.kt

kierse replied:
Huh..ok. So I was just wrong then 😆

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/step/ConfirmationOrderStep.kt:
```diff
@@ -0,0 +1,162 @@
+    val summary = buildSummary(filling)
+    val extra = buildExtras(diet, calories)
+
+    Column(modifier = modifier.fillMaxSize().padding(5.dp)) {
```
these odd-numbered padding values are strange to me. Why not increments of 4/8 like normal?

kierse replied:
No reason. Updated to use increments of 4. https://github.com/slackhq/circuit/pull/496/commits/eee5a45f5db6e43f89110642e8c1955cb202333f

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/step/FillingsOrderStep.kt:
```diff
@@ -0,0 +1,125 @@
+  }
+}
+
+internal class FillingsProducerImpl(private val repository: IngredientsRepository) :
+  OrderStep.StateProducer<FillingsOrderStep.State> {
```
Continuing from my comment higher up - this definitely wants to be a `Presenter`

kierse replied:
https://github.com/slackhq/circuit/pull/496#discussion_r1149871408

ZacSweers commented on samples/tacos/src/main/res/values-night/themes.xml:
```diff
@@ -0,0 +1,31 @@
+
+<resources>
+    <style name="TacoTheme" parent="Theme.Material3.Dark.NoActionBar">
+        <item name="colorPrimary">@color/md_theme_dark_primary</item>
+        <item name="colorOnPrimary">@color/md_theme_dark_onPrimary</item>
```
These xml colors and themes aren't needed right?

ZacSweers replied:
yeah I think since it's all compose we shouldn't need to do much in the xml layer 🤞

kierse commented on samples/tacos/README.md:
```diff
@@ -0,0 +1,37 @@
+There are 2 primary roles:
+
+### Parent circuit
+[OrderTacosCircuit](src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt) maintains order state and manages wizard flow. It:
```
I knew Markdown supported relative paths, but GH support was news to me as well

kierse commented on samples/README.md:
```diff
@@ -10,3 +10,5 @@ See the README.md files in each sample directory for more information.
   use Circuit in a real-world app. Contains lots of cute dogs.
 - `counter` – A multiplatform counter circuit. This circuit is a simple counter that increments or
   decrements a count.
+- `tacos` - A food ordering flow/wizard app demonstrating use of a compose Circuit presenter and
```
Fixed https://github.com/slackhq/circuit/pull/496/commits/986a04b641d0753750d9a0822f1d02b1138d9bce

kierse commented on samples/tacos/README.md:
```diff
@@ -0,0 +1,37 @@
+### Child order step
+Child screens consist of 3 sub-components:
+1. [OrderStep][orderStep] - sealed interface defined by the parent and used when changing wizard steps.
+2. [StateProducer][orderStep] - similar to a Circuit presenter. Implementing class/function accepts order state and an event sink from the parent, and
```
I moved all the links to the bottom of the file and added a new link for StateProducer. https://github.com/slackhq/circuit/pull/496/commits/ca7606c3db4a4d9940f4e44d834158a414d88ac2

ZacSweers commented on samples/tacos/README.md:
Nice cleanups 👍

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
Thanks for doing this in a separate commit, much easier to review 👍

ZacSweers commented on samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt:
```diff
@@ -201,6 +202,9 @@ private fun calculateIngredientsCost(filling: Ingredient?, toppings: Set<Ingredi
 internal fun OrderTacosUi(state: OrderTacosScreen.State, modifier: Modifier = Modifier) {
   val sink = state.eventSink
+
+  BackHandler(enabled = state.isPreviousVisible) { sink(OrderTacosScreen.Event.Previous) }
```
👍

ZacSweers reviewed (approved):
Alright let's roll with this! Looks like just CI that needs taking care of

ZacSweers reviewed (approved):
Remember approvals are sticky in this repo!

ZacSweers commented:
I think you need to merge latest main first as there's a fix for queue concurrency
